snuba/query/snql/parser.py

```diff
@@ -895,6 +901,108 @@ def parse(exp: Expression) -> Expression:
     query.transform_expressions(parse)
 
 
+def _transform_array_condition(array_columns: Set[str], exp: Expression) -> Expression:
+    if not is_condition(exp) or not isinstance(exp, FunctionCall):
+        return exp
+    elif len(exp.parameters) < 2:
+        return exp
+
+    lhs = exp.parameters[0]
+    if not isinstance(lhs, Column):
+        return exp
+
+    aliased_name = (
+        f"{lhs.table_name + '.' if lhs.table_name is not None else ''}{lhs.column_name}"
+    )
+    if aliased_name not in array_columns:
+        return exp
+
+    function_name = (
+        "arrayExists"
+        if FUNCTION_TO_OPERATOR[exp.function_name] in POSITIVE_OPERATORS
+        else "arrayAll"
+    )
+
+    # This is an expression like:
+    # arrayExists(x -> assumeNotNull(notLike(x, rhs)), lhs)
+    return FunctionCall(
+        None,
+        function_name,
+        (
+            Lambda(
+                None,
+                ("x",),
+                FunctionCall(
+                    None,
+                    "assumeNotNull",
+                    (
+                        FunctionCall(
+                            None,
+                            exp.function_name,
+                            (Argument(None, "x"), exp.parameters[1]),
+                        ),
+                    ),
+                ),
+            ),
+            lhs,
+        ),
+    )
+
+
+def _unpack_array_conditions(
+    query: Union[CompositeQuery[QueryEntity], LogicalQuery],
+    schema: ColumnSet,
+    entity_alias: Optional[str] = None,
+) -> None:
+    array_columns = set()
+    array_join_col = query.get_arrayjoin()
+    array_join = ""
+    if array_join_col is not None:
+        assert isinstance(array_join_col, Column)
+        array_join = f"{array_join_col.table_name + '.' if array_join_col.table_name else ''}{array_join_col.column_name}"
+
+    entity_alias = f"{entity_alias}." if entity_alias is not None else ""
+    for column in schema:
+        if isinstance(column.type, Array):
+            aliased_base_name = f"{entity_alias}{column.base_name}"
+            aliased_flattened = f"{entity_alias}{column.flattened}"
```

Comment:
What's the case where we need to take both base and flattened name into account ?

Reply:
There are test cases where the array join is on "exception_stacks" instead of "exception_stacks.<whatever>". This is handling that case. It mirrors the logic that exists in the legacy parser.